### PR TITLE
Add non-generic overloads for A.Fake

### LINKED ISSUE #482: Non-generic Fake and Dummy creation
I just realized that `A.Fake` doesn't have a non-generic overload accepting a `Type` parameter, which would be very useful in some scenarios. I realize that it's not something most people would need on a regular basis, but there doesn't seem to be a good workaround. The only way I found was to call `A.Fake<T>()` via reflection, which is far from ideal.

Am I missing something?

Perhaps adding a non-generic overload to `A.Fake` wouldn't be the best option, because would pollute the API with something that is not usually useful, but it would be nice to have an easy workaround.

### PR DESCRIPTION
**Pull request by thomaslevesque**
and for `A.Dummy`, `A.CollectionOfFake` and `A.CollectionOfDummy`

Fixes #482

### CODE REVIEW

File: tests/FakeItEasy.Specs/NonGenericCreationOptionsSpecs.cs [hunk 1]
```diff
@@ -0,0 +1,845 @@
...
+    using System.Diagnostics.CodeAnalysis;
+    using System.Linq;
+    using System.Reflection.Emit;
+    using FakeItEasy.Creation;
+    using FakeItEasy.Tests;
+    using FakeItEasy.Tests.TestHelpers;
+    using FluentAssertions;
+    using Xbehave;
+
+    public static class NonGenericCreationOptionsSpecs
```
Comment (thomaslevesque) [inline tests/FakeItEasy.Specs/NonGenericCreationOptionsSpecs.cs]: Note: this is an exact copy of `CreationOptionsSpecs`, except that it tests the non-generic `A.Fake`

File: src/FakeItEasy/A.cs [hunk 2]
```diff
@@ -75,6 +107,18 @@ public static T Dummy<T>()
         }
 
         /// <summary>
+        /// Gets a dummy object of the specified type. The value of a dummy object
+        /// should be irrelevant. Dummy objects should not be configured.
+        /// </summary>
+        /// <param name="typeOfDummy">The type of dummy to return.</param>
+        /// <returns>A dummy object of the specified type.</returns>
+        /// <exception cref="ArgumentException">Dummies of the specified type can not be created.</exception>
```
Comment (blairconrad) [inline src/FakeItEasy/A.cs]: It's odd that this method seems to be more accessible than `Dummy<T>` (`EditorBrowsable`). I'd think they should be the same.

File: src/FakeItEasy/A.cs [hunk 3]
```diff
@@ -49,6 +49,27 @@ public static T Fake<T>(Action<IFakeOptions<T>> optionsBuilder)
...
+        {
+            return FakeCreator.CreateFake(typeOfFake, x => { });
+        }
+
+        /// <summary>
+        /// Creates a fake object of the specified type.
+        /// </summary>
+        /// <param name="typeOfFake">The type of fake object to create.</param>
+        /// <param name="optionsBuilder">A lambda where options for the built fake object can be specified.</param>
+        /// <returns></returns>
```
Comment (blairconrad) [inline src/FakeItEasy/A.cs]: Empty `returns`

File: tests/FakeItEasy.Specs/NonGenericCreationOptionsSpecs.cs (hunk 1 above)
Comment (blairconrad) [inline tests/FakeItEasy.Specs/NonGenericCreationOptionsSpecs.cs]: This tempts me to suggest parameterized specs, or a common base, but I wonder if it's worth it.

File: src/FakeItEasy/A.cs (hunk 3 above)
Comment (thomaslevesque) [inline src/FakeItEasy/A.cs]: Oops!

File: src/FakeItEasy/A.cs (hunk 2 above)
Comment (thomaslevesque) [inline src/FakeItEasy/A.cs]: Ah, I missed the `EditorBrowsable` attribute on `Dummy<T>`... I'll add it

File: tests/FakeItEasy.Specs/NonGenericCreationOptionsSpecs.cs (hunk 1 above)
Comment (thomaslevesque) [inline tests/FakeItEasy.Specs/NonGenericCreationOptionsSpecs.cs]: It might be worth it (would avoid a lot of duplication); I'm not sure it could be done, but I'll give it a try.

File: src/FakeItEasy/A.cs (hunk 2 above)
Comment (adamralph) [inline src/FakeItEasy/A.cs]: Shouldn't the same apply to `CollectionOfDummy`?

File: src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs [hunk 4]
```diff
@@ -35,6 +35,14 @@ public T CreateFake<T>(Action<IFakeOptions<T>> optionsBuilder)
             return (T)this.fakeAndDummyManager.CreateFake(typeof(T), options => optionsBuilder((IFakeOptions<T>)options));
         }
 
+        public object CreateFake(Type typeOfFake, Action<IFakeOptions> optionsBuilder)
+        {
+            Guard.AgainstNull(typeOfFake, nameof(typeOfFake));
+            Guard.AgainstNull(optionsBuilder, nameof(optionsBuilder));
```
Comment (adamralph) [inline src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs]: This method does not de-reference either argument. These guards should only be added if it's not possible for a method further down the call stack to give the same quality of exception message. Has that been verified?

File: src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs [hunk 5]
```diff
@@ -55,6 +63,20 @@ public IList<T> CollectionOfFake<T>(int numberOfFakes)
             return result;
         }
 
+        public IList<object> CollectionOfFake(Type typeOfFake, int numberOfFakes)
+        {
+            Guard.AgainstNull(typeOfFake, nameof(typeOfFake));
```
Comment (adamralph) [inline src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs]: Same question again about the guard.

File: src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs [hunk 6]
```diff
@@ -68,6 +90,13 @@ public T CreateDummy<T>()
             return (T)this.fakeAndDummyManager.CreateDummy(typeof(T));
         }
 
+        public object CreateDummy(Type typeOfDummy)
+        {
+            Guard.AgainstNull(typeOfDummy, nameof(typeOfDummy));
```
Comment (adamralph) [inline src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs]: And again.

File: src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs [hunk 7]
```diff
@@ -87,5 +116,19 @@ public IList<T> CollectionOfDummy<T>(int numberOfDummies)
 
             return result;
         }
+
+        public IList<object> CollectionOfDummy(Type typeOfDummy, int numberOfDummies)
+        {
+            Guard.AgainstNull(typeOfDummy, nameof(typeOfDummy));
```
Comment (adamralph) [inline src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs]: And again.

File: src/FakeItEasy/A.cs (hunk 2 above)
Comment (blairconrad) [inline src/FakeItEasy/A.cs]: If we're adding `EditorBrowsable`s, absolutely

Comment (adamralph) [inline src/FakeItEasy/A.cs]: I'd be happy with the missing ones being added in a separate commit in this PR.

**Comment by adamralph** [comment]
LGTM. My comments are minor concerns.

**Comment by blairconrad** [comment]
Me too.

Oh! Docs.

File: src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs (hunk 4 above)
Comment (thomaslevesque) [inline src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs]: Ah, I wasn't aware of this rule. I saw a `Guard.AgainstNull` in the generic `CreateFake<T>`, so I added it here as well.

**Comment by thomaslevesque** [comment]
> Oh! Docs.

Ah, yes. Thanks for the reminder!

File: src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs (hunk 4 above)
Comment (adamralph) [inline src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs]: No problem, fully understandable. I believe we're using null guards in a lot of redundant places. It may be worth doing a cross cutting sweep of all of them at some point and we should consider our approach then.

For this PR, I think it's fine as-is.

File: docs/creating-fakes.md [hunk 8]
```diff
@@ -8,6 +8,17 @@ var foo = A.Fake<IFoo>();
 ```
 This will return a faked object that is an actual instance of the type specified (`IFoo` in this case).
 
+You can also create a collection of fakes by writing:
+```csharp
+var foos = A.CollectionOfFake<Foo>(10);
+```
```
Comment (thomaslevesque) [inline docs/creating-fakes.md]: Not really related to this issue, but I noticed `CollectionOfFake<T>` wasn't introduced. It would have been weird to mention the non-generic one without mentioning the generic one, so I added it as well.

File: docs/creating-fakes.md [hunk 9]
```diff
@@ -31,14 +42,14 @@ var foo = A.Fake<FooClass>(x => x.WithArgumentsForConstructor(() => new FooClass
 var foo = A.Fake<FooClass>(x => x.WithArgumentsForConstructor(new object[] { "foo", "bar" }));
 
 // Specifying additional interfaces to be implemented. Among other uses,
-// this can help when a fake skips members because they have been
```
Comment (thomaslevesque) [inline docs/creating-fakes.md]: editorconfig

File: src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs [hunk 10]
```diff
@@ -35,6 +35,14 @@ public T CreateFake<T>(Action<IFakeOptions<T>> optionsBuilder)
             return (T)this.fakeAndDummyManager.CreateFake(typeof(T), options => optionsBuilder((IFakeOptions<T>)options));
         }
 
+        public object CreateFake(Type typeOfFake, Action<IFakeOptions> optionsBuilder)
```
Comment (thomaslevesque) [inline src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs]: I realized I didn't copy the documentation from the interface. It's an internal class, so I'm not sure it really matters, but if you feel it's important, I'll add it.

File: src/FakeItEasy/Creation/IFakeCreatorFacade.cs [hunk 11]
```diff
@@ -18,8 +18,17 @@ internal interface IFakeCreatorFacade
...
+        /// <param name="typeOfFake">The type of fake to create.</param>
+        /// <param name="optionsBuilder">Action that builds options for the created fake object.</param>
+        /// <returns>The created fake object.</returns>
+        /// <exception cref="FakeItEasy.Core.FakeCreationException">Was unable to generate the fake in the current configuration.</exception>
+        object CreateFake(Type typeOfFake, Action<IFakeOptions> optionsBuilder);
+
+        /// <summary>
         /// Creates a dummy object, this can be a fake object or an object resolved
-        /// from the current IFakeObjectContainer.
+        /// from a dummy factory.
```
Comment (thomaslevesque) [inline src/FakeItEasy/Creation/IFakeCreatorFacade.cs]: There seems to be a few places where `IFakeObjectContainer` is still mentioned... we should probably make a sweep to fix it everywhere.

File: tests/FakeItEasy.Specs/CreationOptionsSpecs.cs [hunk 12]
```diff
@@ -835,12 +835,30 @@ public interface IInterfaceThatWeWillAddAttributesTo3
...
             "When I create a fake using the options builder"
-                .x(() => exception = Record.Exception(() => A.Fake(optionsBuilder)));
+                .x(() => exception = Record.Exception(() => this.CreateFake(optionsBuilder)));
 
             "Then fake creation fails"
                 .x(() => exception.Should().BeAnExceptionOfType<ArgumentException>()
                 .WithMessage("*Only expression of the type ExpressionType.New (constructor calls) are accepted.*"));
         }
+
+        protected abstract T CreateFake<T>(Action<IFakeOptions<T>> optionsBuilder);
```
Comment (thomaslevesque) [inline tests/FakeItEasy.Specs/CreationOptionsSpecs.cs]: This is the simplest way I found to test both the generic and non-generic overloads. Do you see any issue with it?

File: tests/FakeItEasy.Specs/DummyCreationSpecs.cs [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using FakeItEasy.Tests.TestHelpers;
+    using FluentAssertions;
+    using Xbehave;
+
+    public abstract class DummyCreationSpecsBase
```
Comment (thomaslevesque) [inline tests/FakeItEasy.Specs/DummyCreationSpecs.cs]: There were no specs for dummy creation, so I added some. Probably partially redundant with `DummyTests`

File: src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs (hunk 10 above)
Comment (adamralph) [inline src/FakeItEasy/Creation/DefaultFakeCreatorFacade.cs]: I don't think we need it.

File: src/FakeItEasy/Creation/IFakeCreatorFacade.cs (hunk 11 above)
Comment (adamralph) [inline src/FakeItEasy/Creation/IFakeCreatorFacade.cs]: Raised https://github.com/FakeItEasy/FakeItEasy/issues/757

File: src/FakeItEasy/A.cs [hunk 14]
```diff
@@ -113,6 +113,7 @@ public static T Dummy<T>()
         /// <param name="typeOfDummy">The type of dummy to return.</param>
         /// <returns>A dummy object of the specified type.</returns>
         /// <exception cref="ArgumentException">Dummies of the specified type can not be created.</exception>
+        [EditorBrowsable(EditorBrowsableState.Advanced)]
```
Comment (adamralph) [inline src/FakeItEasy/A.cs]: This should really have been added in the commit which added this new overload, but I'm happy to proceed as-is.

(Same applies to `CollectionOfDummy(Type typeOfDummy, int numberOfDummies)`.)

**Comment by adamralph** [comment]
LGTM - I'm :+1: to merge.

File: src/FakeItEasy/A.cs (hunk 14 above)
Comment (thomaslevesque) [inline src/FakeItEasy/A.cs]: At some point you said you wanted all `EditorBrowsable` attributes to be added in a separate commit; maybe I misunderstood. Anyway, I don't mind squashing this commit if you prefer.

Comment (adamralph) [inline src/FakeItEasy/A.cs]: Sorry, I wasn't clear. I meant only the ones that are currently missing in master (there's only one).

Anyway, I'm happy as-is. No need to squash.

File: src/FakeItEasy/Creation/IFakeCreatorFacade.cs [hunk 15]
```diff
@@ -28,6 +37,16 @@ internal interface IFakeCreatorFacade
         T CreateDummy<T>();
 
         /// <summary>
+        /// Creates a dummy object, this can be a fake object or an object resolved
+        /// from a dummy factory.
+        /// </summary>
+        /// <param name="typeOfDummy">The type of dummy to create.</param>
+        /// <returns>The created dummy.</returns>
+        /// <exception cref="FakeItEasy.Core.FakeCreationException">Was unable to generate the fake in the current configuration and
+        /// no dummy was registered in the container for the specified type..</exception>
```
Comment (blairconrad) [inline src/FakeItEasy/Creation/IFakeCreatorFacade.cs]: The double "." is odd here, but the docs were just copied, so not your fault. At least it's consistent.

File: tests/FakeItEasy.Specs/CreationOptionsSpecs.cs (hunk 12 above)
Comment (blairconrad) [inline tests/FakeItEasy.Specs/CreationOptionsSpecs.cs]: This is about what I imagined happening. An alternative would be to split the two concrete classes into their own files, but everything's so tightly related that I don't mind it as is.

File: tests/FakeItEasy.Specs/DummyCreationSpecs.cs (hunk 13 above)
Comment (blairconrad) [inline tests/FakeItEasy.Specs/DummyCreationSpecs.cs]: Commendable! I see no need to root out the redundancies.

Comment (blairconrad) [inline tests/FakeItEasy.Specs/DummyCreationSpecs.cs]: … but the day the redundant unit tests inconvenience me: :skull:

File: src/FakeItEasy/Creation/IFakeCreatorFacade.cs (hunk 15 above)
Comment (blairconrad) [inline src/FakeItEasy/Creation/IFakeCreatorFacade.cs]: We could leave and clean up in another PR.

Comment (thomaslevesque) [inline src/FakeItEasy/Creation/IFakeCreatorFacade.cs]: Fixed

**Comment by blairconrad** [comment]
As usual, sterling work, @thomaslevesque. I did make a minor note about (mostly existing) XML docs, but ~~I think we can wait~~ you corrected it while I was typing this note!

**Comment by thomaslevesque** [comment]
Thanks!

**Comment by adamralph** [comment]
:+1: great work all round!

BTW - don't forget to add the closed issue to the next milestone. I've done it for #482

**Comment by thomaslevesque** [comment]
> BTW - don't forget to add the closed issue to the next milestone.

will do

**Comment by blairconrad** [comment]
> BTW - don't forget to add the closed issue to the next milestone.

As merger, I should've done this. :blush:
